### PR TITLE
Concatenate any state

### PR DESCRIPTION
**Pull request by sduquemesa**
**Context:**
Concatenation of non-gaussian states was not yet implemented in Mr Mustard.

**Description of the Change:**
This PR implements concatenation of non-gaussian states. To do so, kets for pure states represented with a density matrix can now be calculated using the new implemented `fock.dm_to_ket` function.

**Benefits:**
Have the ability to apply a circuit to one or more non-Gaussian inputs.

**Possible Drawbacks:**
Once concatenation is performed, all calculations in Mr Mustard will proceed in the Fock representation hence having an impact on performance. The precision to which the new concatenated states represents the original states depends on the assigned cutoffs — a better representation requires higher cutoff values.

**Related GitHub Issues:**
None

### CODE REVIEW

File: mrmustard/lab/abstract/state.py [hunk 1]
```diff
@@ -390,11 +396,35 @@ def primal(self, other: Union[State, Transformation]) -> State:
...
         r"""Concatenates two states."""
-        if not (self.is_gaussian and other.is_gaussian):
-            raise NotImplementedError(
-                "Concatenation of non-gaussian states is not implemented yet."
-            )
-
+        if not self.is_gaussian or not other.is_gaussian:  # convert all to fock now
+            if (
+                self.is_mixed or other.is_mixed
+            ):  # TODO: would be more efficient if we could keep pure states as kets
```
Comment (thisac) [inline mrmustard/lab/abstract/state.py:402]: (minor cosmetic suggestion)
```suggestion
            # TODO: would be more efficient if we could keep pure states as kets
            if self.is_mixed or other.is_mixed:
```

File: mrmustard/physics/fock.py [hunk 2]
```diff
@@ -123,6 +125,37 @@ def ket_to_dm(ket: Tensor) -> Tensor:
...
 
 
+def dm_to_ket(dm: Tensor) -> Tensor:
+    r"""Maps a density matrix to a ket if the state is pure.
+
+    If the state is pure :math:`\hat \rho= |\psi\rangle\langle \psi|` then the
+    ket is the eigenvector of ``\rho`` corresponding to the eigenvalue 1.
+
+    Args:
+        dm: the density matrix
```
Comment (thisac) [inline mrmustard/physics/fock.py:135]:
```suggestion
        dm (Tensor): the density matrix
```

File: mrmustard/physics/fock.py [hunk 3]
```diff
@@ -123,6 +125,37 @@ def ket_to_dm(ket: Tensor) -> Tensor:
     return math.outer(ket, math.conj(ket))
 
 
+def dm_to_ket(dm: Tensor) -> Tensor:
+    r"""Maps a density matrix to a ket if the state is pure.
+
+    If the state is pure :math:`\hat \rho= |\psi\rangle\langle \psi|` then the
+    ket is the eigenvector of ``\rho`` corresponding to the eigenvalue 1.
```
Comment (thisac) [inline mrmustard/physics/fock.py:132]: Should be `:math:`, no?
```suggestion
    ket is the eigenvector of :math:`\rho` corresponding to the eigenvalue 1.
```

File: tests/test_lab/test_states.py [hunk 4]
```diff
@@ -191,3 +191,11 @@ def test_getitem_set_modes(modes):
...
 
     assert state1.modes == state2.modes
+
+
+def test_concat_fock_gaussian_states():
+    Ga = Gaussian(1)
+    Gb = Gaussian(1)
+    a = State(dm=Ga.dm())
+    b = State(dm=Gb.dm())
+    a & Gb == Ga & b
```
Comment (thisac) [inline tests/test_lab/test_states.py:201]: This test doesn't seem to assert anything or catch any errors? 🤔

File: mrmustard/physics/fock.py [hunk 5]
```diff
@@ -123,6 +125,37 @@ def ket_to_dm(ket: Tensor) -> Tensor:
...
+
+    Args:
+        dm: the density matrix
+
+    Returns:
+        Tensor: the ket
+    """
```
Comment (thisac) [inline mrmustard/physics/fock.py:139]: (minor) missing raises
```suggestion
    Returns:
        Tensor: the ket
        
    Raises:
        ValueError: if ket for mixed states cannot be calculated
    """
```

**Comment by sduquemesa** [comment]
Thanks @thisac for your ultra-quick review! This is still work in progress - I forgot to put the WIP label 🥲

**Comment by thisac** [comment]
> Thanks @thisac for your ultra-quick review! This is still work in progress - I forgot to put the WIP label 

Oh, sorry about that. Just had a quick look and noticed some things, missed that it was still a draft. Didn't do a full review of it. 😬

File: tests/test_lab/test_states.py [hunk 6]
```diff
@@ -191,3 +210,28 @@ def test_getitem_set_modes(modes):
...
+
+    if not pure:
+        state1 >>= Attenuator(transmissivity=0.95)
+        state2 >>= Attenuator(transmissivity=0.9)
+
+    psi = state1 & state2
+
+    # test concatenated state
+    psi_dm = math.transpose(math.tensordot(state1.dm(), state2.dm(), [[], []]), [0, 2, 1, 3])
+    assert np.allclose(psi.dm(), psi_dm, rtol=1e-2)
```
Comment (ziofil) [inline tests/test_lab/test_states.py:229]: is the high relative tolerance necessary?

**Review by ziofil — APPROVED** [review]
Super!

File: tests/test_lab/test_states.py (hunk 6 above)
Comment (sduquemesa) [reply tests/test_lab/test_states.py:229]: Not actually. I was testing something and forgot to update this. Thanks for spotting it!

**Pull request merged**